src/rgw/rgw_dedup_cluster.cc

```diff
@@ -0,0 +1,1068 @@
+	return -EINVAL;
+      }
+      if (ret != 0) {
+	ldpp_dout(dpp, 1) << __func__ << "::Bad notify ack, ret=" << ret
+			  << "::err=" << cpp_strerror(ret) << dendl;
+	return ret;
+      }
+    }
+    ldpp_dout(dpp, 10) << __func__ << "::" << get_urgent_msg_names(urgent_msg)
+		       << " finished successfully!" << dendl;
+    return 0;
+  }
+
+  //---------------------------------------------------------------------------
+  // command-line called from radosgw-admin.cc
+  int cluster::dedup_restart_scan(rgw::sal::RadosStore *store,
+				  bool dry_run,
+				  const DoutPrefixProvider *dpp)
+  {
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      return -1;
+    }
+
+    dedup_epoch_t old_epoch;
+    // store the previous epoch for cmp-swap
+    int ret = get_epoch(ioctx, dpp, &old_epoch, __func__);
+    if (ret != 0) {
+      return ret;
+    }
+
+    // first abort all dedup work!
+    ret = dedup_control(store, dpp, URGENT_MSG_ABORT);
+    if (ret != 0) {
+      return ret;
+    }
+
+    ldpp_dout(dpp, 1) << __func__ << (dry_run ?"::DRY RUN":"::FULL DEDUP") << dendl;
+    int dedup_type = (dry_run ? DEDUP_TYPE_DRY_RUN : DEDUP_TYPE_FULL);
+    ret = swap_epoch(dpp, ioctx, &old_epoch, dedup_type, 0, 0);
+    if (ret == 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::Epoch object was reset" << dendl;
+      dedup_control(store, dpp, URGENT_MSG_RESTART);
+    }
+    else {
+      ldpp_dout(dpp, 0) << __func__ << "::Failed to reset EPOCH object with: "
+			<< cpp_strerror(ret) << ", ret=" << ret << dendl;
+    }
+    return ret;
+  }
+
+  //---------------------------------------------------------------------------
+  bool cluster::can_start_new_scan(rgw::sal::RadosStore *store)
+  {
+    ldpp_dout(dpp, 10) << __func__ << "::epoch=" << d_epoch_time << dendl;
+    librados::IoCtx ioctx;
+    if (init_dedup_pool_ioctx(store->getRados(), dpp, ioctx) != 0) {
+      return -1;
+    }
+
+    dedup_epoch_t new_epoch;
+    if (get_epoch(ioctx, dpp, &new_epoch, nullptr) != 0) {
+      ldpp_dout(dpp, 1) << __func__ << "::No Epoch Object::"
+			<< "::scan can be restarted!\n\n\n" << dendl;
+      // no epoch object exists -> we should start a new scan
+      return true;
+    }
+
+    if (new_epoch.time <= d_epoch_time) {
+      if (new_epoch.time == d_epoch_time) {
+	ldpp_dout(dpp, 10) << __func__ << "::Epoch hasn't change - > Do not restart scan!!" << dendl;
+      }
+      else {
+	ldpp_dout(dpp, 1) << __func__ << "::Do not restart scan!\n    epoch="
+			  << d_epoch_time << "\nnew_epoch="<< new_epoch.time <<dendl;
+      }
+      return false;
+    }
+    // allow members to join within a 30 sec limit
+    utime_t limit = {30, 0};
+    utime_t now = ceph_clock_now();
+    ldpp_dout(dpp, 1) << __func__ << "\n::new_epoch=" << new_epoch.time
+		      << "\n::now      =" << now << dendl;
+    if ((now > new_epoch.time) && ((now - new_epoch.time) < limit)) {
```

Comment:
what about the case that `now > epoch` but the diff is more than 30 sec?

Reply:
Members can only join the scan within a 30 seconds window.
If they arrive after more than 30 seconds they are ineligible to join the scan.

I will need to review this part of the code as it has heavy implications.
Leaving this item unresolved for now